packages/blaze-query/src/hydra.ts

```diff
@@ -0,0 +1,758 @@
+      monetaryExpansion: rawParams.monetaryExpansion,
+      treasuryExpansion: rawParams.treasuryCut,
+      minPoolCost: rawParams.minPoolCost,
+      protocolVersion: rawParams.protocolVersion,
+      maxExecutionUnitsPerBlock: {
+        memory: rawParams.maxBlockExecutionUnits.memory,
+        steps: rawParams.maxBlockExecutionUnits.steps,
+      },
+      maxBlockBodySize: rawParams.maxBlockBodySize,
+      maxBlockHeaderSize: rawParams.maxBlockHeaderSize,
+    };
+  }
+
+  async getUnspentOutputs(
+    address: Address,
+  ): Promise<TransactionUnspentOutput[]> {
+    const utxos = await this.fetchUTxOs();
+
+    const ret: TransactionUnspentOutput[] = [];
+    for (const key in utxos) {
+      const utxo = utxos[key]!;
+      if (utxo.output().address().toBech32() === address.toBech32()) {
+        ret.push(utxo);
+      }
+    }
+
+    return Promise.resolve(ret);
+  }
+
+  async getUnspentOutputsWithAsset(
+    address: Address,
+    unit: AssetId,
+  ): Promise<TransactionUnspentOutput[]> {
+    const utxos = await this.fetchUTxOs();
+
+    const ret: TransactionUnspentOutput[] = [];
+    for (const key in utxos) {
+      const utxo = utxos[key]!;
+      if (utxo.output().amount().multiasset()?.has(unit)) {
+        ret.push(utxo);
+      }
+    }
+
+    return Promise.resolve(ret);
+  }
+
+  /**
+   * This method fetches the UTxO that holds a particular NFT given as an argument
+   * It does **NOT** gaurantee the uniqueness of the NFT.
+   * It is just like `getUnspentOutputsWithAsset`, except it short circuits and returns the first UTxO containing the unit
+   * @param unit - the AssetId of the NFT
+   * @returns A promise that resolves to the UTxO
+   */
+  async getUnspentOutputByNFT(
+    unit: AssetId,
+  ): Promise<TransactionUnspentOutput> {
+    const utxos = await this.fetchUTxOs();
+
+    for (const key in utxos) {
+      const utxo = utxos[key]!;
+      if (utxo.output().amount().multiasset()?.has(unit)) {
+        return utxo;
+      }
+    }
+
+    throw new Error("UTxO not found");
+  }
+
+  async resolveUnspentOutputs(
+    txIns: TransactionInput[],
+  ): Promise<TransactionUnspentOutput[]> {
+    const utxos = await this.fetchUTxOs();
+
+    const ret: TransactionUnspentOutput[] = [];
+    for (const txIn of txIns) {
+      const utxo: TransactionUnspentOutput | undefined =
+        utxos[`${txIn.transactionId()}#${txIn.index()}`];
+      if (!utxo) {
+        throw new Error(
+          `Failed to resolve transaction: ${txIn.transactionId()}#${txIn.index()}`,
+        );
+      }
+
+      ret.push(utxo);
+    }
+
+    return ret;
+  }
+
+  /**
+   * This method takes a datum hash and tries to find the assosciated inline datum in the current snapshot utxo set.
+   * If the output containing the datum has been spent, this method will error
+   * @param datumHash
+   * @returns Promise<PlutusData>
+   */
+  async resolveDatum(datumHash: DatumHash): Promise<PlutusData> {
```

Comment:
Is this the same behavior as blaze would expect from other providers? AFAIK its often full-fledged indexers that we have providers for and those would probably index datums of all transaction bodies (as auxiliary or witness).

Reply:
Providers are abstractly nodes with indexers (which is a design inherited from lucid). E.g ogmios alone can't be a provider (this may have changed if ogmios has updated) because it can't resolve datums, so you have Kupmios. Every major hosted service includes indexation. I'm surprised the node doesn't support datum resolution natively.

Reply:
Yeah, that is one reason my hydra provider doesn't quite line up with the concept of providers in blaze largely. However, If we want to use blaze to build and submit transactions to hydra, I do have to have a hydra provider to communicate with the node.

Perhaps we need to break up a provider into smaller pieces? @micahkendall

Reply:
FWIW kupo does support Hydra https://cardanosolutions.github.io/kupo/#section/Getting-started/-hydra-host-hostname-hydra-port-port-number

So we could be using kupo + transaction submission directly into Hydra?